### PR TITLE
[-] BO : Speed up search indexing, reduce memory usage and fix attribute indexing (PSCSX-5039, PSCSX-5028)

### PR DESCRIPTION
**Pull request by jocel1**
On a DB with 1925 products, for a full reindex :

Before this fix : 
- 12.8 Mo peak memory consumption
- 101s to index the products

After this fix :
- 8.1 Mo peak memory consumption
- 56s to index the products

### CODE REVIEW

**Comment by doekia** [comment]
Got the feeling there is a mistake.

We should run the update with the join before the delete otherwise the update will do nothing or cannot filter on what we have deleted.

Don't you think so?

**Comment by jocel1** [comment]
Actually in the last commit I removed the table search_index from the join. For me, if $id_product is not specified, this update is only here to be sure product_shop and product have the same value in the "indexed" field.

**Comment by eric-rouvier** [comment]
Thanks Jocelyn. Now it's more clear, so some remarks raise... (Search.php)

First, on line 606 you kept the limit to 50. It was the actual max number of entries to be processed at each step. Now it is the number of products whose entries will be processed: nof_prods * nof_attributes * nof_languages. Can be much more than 50...

Then, not important: in block 561-575, I think that $id_product is not false (so doesn't need to be tested) except if both $full and $id_product are false. In that case we can return immediately, can't we? By the way, aren't the two arguments of indexation redundant? (method expects $full XOR $id_product)

Finally, a question. I'm not a DB specialist, but when I have to implement such a function, I would query once and fetch row by row, hoping the DB engine to be intelligent enough not to get out of memory. I never looked deep into PS DB implementation, but I understood from your discussion with doekia, that there are several ones. Would it be difficult to implement an internal buffer maintaining offset and limit in order to provide this facility even if the DB engine does not? I don't know how the different Db instances can coexist peacefully...

**Comment by doekia** [comment]
@eric-rouvier 

Your point is valid regarding row by row, however modern install use DbPDO. At creation of the DbPDO, we use PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => true. This mean that the entire recordset is fetched inside the memory (counted in PHP memory use). Not doing such could lead to issues with multiple intricated query() not returnning the entire set.
I have proposed a patch for cursors https://github.com/PrestaShop/PrestaShop/pull/2572, even do it works on every implementation I have tested this could lead to some other issues with DbDBO::numRows() not returning the proper count of record as stated in some pdo bug report. The patch is actually under investigation.

IMHO, some settings.inc.php constant should let the sysadmin decide which db abstraction to use, so small (default) shop will use DbPDO and medium to big shop could enforce DbMySQLi ...

**Comment by jocel1** [comment]
I don't think keeping 50 as a limit is an issue. Even with attributes & languages, the final number will be pretty low (on my local test, I didn't see any issue with it).

As for the full vs id_product, you have 3 possibilities : provide an id_product or do not provide it and instead use full = false (index not indexed yet products) or full = true (full reindex). So the case false, false is a perfectly valid combination.

As for the DB, on line 443 you can see I've set the "array" parameter to false, to avoid fetching three times the datas into an array.

**Comment by jocel1** [comment]
@doekia : just curious, have you tried my patch on your datas (PSCSX-5039) (I wonder what could be the speed & memory consumption in your case)

**Comment by eric-rouvier** [comment]
OK guys, got it. From a practical point of view, should I close my initial PR #2651?

**Comment by doekia** [comment]
@jocel1 shame on me but I don't know how to apply this single commit to my branch ? any pointer?

**Comment by guestisp** [comment]
@jocel1 keep in mind that we have many shops with > 25.000 products and more than 50 attributes in 5 languages, thus 50 as limit could be low if calc is nof_prods * nof_attributes * nof_languages.

**Comment by jocel1** [comment]
@guestisp : the new implementation could actually process more product than before, so it should be ok (it's now 50 products, will all the languages and all the attributes for each products)

**Comment by jocel1** [comment]
@doekia : you could simply try to apply the associated patch : https://github.com/PrestaShop/PrestaShop/pull/2649.patch

**Comment by doekia** [comment]
@jocel1 
Ok let's try to be synthetic.

On the machine that initially triggered the 64GB issue.
PHP 5.4.38-1~dotdeb.1 (cli) (built: Feb 19 2015 18:54:29) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2014 Zend Technologies
    with XCache v2.0.0, Copyright (c) 2005-2012, by mOo
    with the ionCube PHP Loader v4.7.5, Copyright (c) 2002-2014, by ionCube Ltd.
PDO Driver for MySQL => enabled
Client API version => mysqlnd 5.0.10 - 20111026 - $Id: c85105d7c6f7d70d609bb4c000257868a40840ab $

The patch has been applied on a pristine 1.6.0.14 and gives the following results:
9699328/9699328   0 products
20447232/20709376 100 products
90701824/90963968  1000 products
172228608/172490752 2000 products
memory usage / peak memory usage

I tries on 2 differents system to apply the patch over a 1.6.1.0.
PHP 5.5.9-1ubuntu4.6 (cli) (built: Feb 13 2015 19:18:15) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with the ionCube PHP Loader v4.7.3, Copyright (c) 2002-2014, by ionCube Ltd., and
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies
PDO Driver for MySQL => enabled
Client API version => 5.5.41

PHP 5.5.9-1ubuntu4.6 (cli) (built: Feb 13 2015 19:18:15) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with the ionCube PHP Loader v4.7.3, Copyright (c) 2002-2014, by ionCube Ltd., and
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies
PDO Driver for MySQL => enabled
Client API version => mysqlnd 5.0.11-dev - 20120503 - $Id: bf9ad53b11c9a57efdb1057292d73b928b8c5c77 $

Both system poorly failed by blowing up on memory during DbPDO::getAll()
They have not even proceed a single record.


Hope this helps because I cannot re run those tests again due lack of time

**Comment by jocel1** [comment]
Do you which getAll() ?

**Comment by doekia** [comment]
@jocel1
forgive me. On the last 2 system since I installed thru git I have run git apply --stat 2572.patch
which for whatever reason did not patch the files!!
So the all the systems behave similarly.

Sorry for the call - sound like we have a full running solution ;-)

**Comment by jocel1** [comment]
That's a really good news :)

File: classes/Search.php [hunk 1]
```diff
@@ -397,12 +397,58 @@ public static function getFeatures($db, $id_product, $id_lang)
...
+				WHERE product_shop.`indexed` = 0
+				AND product_shop.`visibility` IN ("both", "search")
+				AND product_shop.`active` = 1
+				ORDER BY product_shop.`id_product` ASC
+				LIMIT '.(int)$limit;
+
+
+			$res = Db::getInstance()->executeS($sql, false);
+			while($row = Db::getInstance()->nextRow($res))
+			{
```
Comment (jeromenadaud) [inline classes/Search.php]: norms ! ;-)

**Comment by WisienDot** [comment]
Same at my shop - index builded successfully but no items in the search result. I decided to use this Search class only to rebuild my index and after that switched classes again - works till I find a solution.

**Pull request merged**